### ISSUE TITLE
Requests with #Points > 2, duplicated points at waypoints

### ISSUE THREAD

**Issue by boldtrn** [graphhopper/graphhopper#1139]
I just understood the issue with the extra point for requests which have via points, that we discussed [here](https://github.com/graphhopper/graphhopper/pull/1091#issuecomment-321506225).

When there is a GHRequest with for example 3 Points. We merge two Paths. Usually, the last point of the first Path is equal to the first point of the second Path. Right now we just append the two Paths. This leads to a duplication of points. For example have a look at this [route](https://graphhopper.com/maps/?point=49.984352%2C11.498802&point=49.984565%2C11.499188&point=49.9847%2C11.499612&points_encoded=false).

The resulting points json looks like this:
```
points: {
coordinates: [
[
11.498807,
49.984347
],
[
11.499067,
49.98446
],
[
11.499214,
49.984541
],
=============================
Here is the waypoint, you see that the point above and below are duplicated
=============================
[
11.499214,
49.984541
],
[
11.499627,
49.984682
]
],
type: "LineString"
},
```
Overall, this is not very critical. The better or cleaner solution would probably be to not duplicate these points, but we would have to change our instructions then.

Right now the ViaPointInstruction, references the first duplicated point. The ContinueInstruction references the second duplicated point. If we remove one of these points, we could not continue with this pattern.

```
{
distance: 0,
sign: 5,
interval: [
2,
3
],
text: "Waypoint 1",
time: 0,
street_name: ""
},
{
distance: 33.387,
sign: 0,
interval: [
3,
4
],
text: "Continue onto Waldhüttenstraße",
time: 7512,
street_name: "Waldhüttenstraße"
},
```

**Comment by karussell** [comment]
> but we would have to change our instructions then.

Yes, I would postpone this when we make instructions a new PathDetail? Otherwise this feels a bit artificial if we skip all via points that have just one point but the FinishInstruction will be added as all instructions do not include the last point ...

**Comment by boldtrn** [comment]
Let's discuss this here instead of in #1138 :).

> [Hmmh, this feels wrong and we should probably skip adding the via points opposed to my recent comment in #1139?](https://github.com/graphhopper/graphhopper/pull/1138#discussion_r133962437)

What do you mean? Should we change the instructions then?

We could remove the first point of the first instruction, for every Path, except the first Path. But what happens if the first instruction after the via-point has only one point?

**Comment by karussell** [comment]
> What do you mean? Should we change the instructions then?

Can we avoid adding points from a ViaInstruction to the "fullPoins" in PathMerger?

Something like:

```java
if(!(i instanceof FinishInstruction) || pathIndex + 1 == paths.size())
   fullPoints.add(i.getPoints());
```

**Comment by karussell** [comment]
Or should we make the correct refactoring? Currently the normal instructions have the interval `[from, last-1]` and the ViaInstructions and FinishInstructions have `[from,last]`. The interval `[from,last]` should be the default but then we should ignore all instructions with last-from<1. This would then also be in sync with the PathDetails that also have `[first,last]`

**Comment by boldtrn** [comment]
> Can we avoid adding points from a ViaInstruction to the "fullPoins" in PathMerger?
> Or should we make the correct but Java-API-breaking refactoring?

I think that, your described change would not fix the issue. The issue is that the ViaInstruction and the next instruction after the ViaInstruction reference the same coordinate.

Currently, every point in the points array, belongs to exactly one Instruction. So the "correct" change would be to break this assumption. To use the example given above, we would remove the duplicated point

```
[
11.499214,
49.984541
]
```

Then we would have to change the references of the Instructions to:

```
{
distance: 0,
sign: 5,
interval: [
2,
3
],
text: "Waypoint 1",
time: 0,
street_name: ""
},
{
distance: 33.387,
sign: 0,
interval: [
2,
3
],
text: "Continue onto Waldhüttenstraße",
time: 7512,
street_name: "Waldhüttenstraße"
},
```

> Currently the normal instructions have the interval [from, last-1] and the ViaInstructions and FinishInstructions have [from,last]

Are you sure? To me it looks like the other way around.

In [InstructionList#createJson](https://github.com/graphhopper/graphhopper/blob/91ca93d8520480f3d103a79962ecabad35f72c78/core/src/main/java/com/graphhopper/util/InstructionList.java#L107) we do:

```
// the last instruction should not point to the next instruction
if (counter + 1 == instructions.size())
   tmpIndex--;
```

**Comment by karussell** [comment]
I think (95% sure ;)) the current state for two paths (size 10 and 10) with a via point is:

[0, 9], [10, 10], [11, 20]

but it should be:

[0, 10], [10, 10], [10, 19]

where the point at index 10 is the end point of the first **and** the via point **and** the start point of the second path but it should not be duplicated in the final PointList as it is now. This way the instructions would be in-line with the PathDetails.

**Comment by boldtrn** [comment]
For example look at this [request](https://graphhopper.com/api/1/route?point=49.984352%2C11.498802&point=49.984565%2C11.499188&point=49.98481%2C11.500009&type=json&locale=en-AU&vehicle=car&weighting=fastest&elevation=true&key=2a24e316-61ea-4850-b231-4ef2fe25d229&points_encoded=false). You can see that the 9,9 example is true for the finish instruction, but not the via instruction. For the via instruction it would be 8,9.

**Comment by karussell** [comment]
> You can see that the 9,9 example is true for the finish instruction, but not the via instruction. For the via instruction it would be 8,9.

Yes, it is inconsistent. But isn't it [9,10]?

So, I think the current state is: [0, 9], via:[9, 10], [10, 19], finish:[19, 19]

Which does not make much sense as a via point should be just one point.
So it should be: [0, 9], via:[9, 9], [9, 18], finish:[18, 18]

A further problem is also on the Java side: here we measure a length of a ViaInstruction with the help of PointList.size which would be 1 instead of last-first which would be 0.

**Comment by devemux86** [comment]
> Which does not make much sense as a via point should be just one point.
And it should be: [0, 9], via:[9, 9], [9, 18], finish:[18, 18]

Indeed that's more logical.

**Comment by devemux86** [comment]
I think seeing different behavior between clients: Java vs Web.
Java client returns open [x,y) instruction intervals, while Web client returns closed [x,y] ones?

e.g. for a small path with 2 instructions and `pointList.size() = 2`, iterating `instruction.getPoints().size()` returns 1+1 points in Java and 2+1 points in Web.

**Comment by boldtrn** [comment]
> Java client returns open [x,y) instruction intervals, while Web client returns closed [x,y] ones?

With Java client you mean that you include GraphHopper as a lib and call the calcPaths method locally? With web you mean the json interface or do you mean the java client?

The Java interface does not really provide you any intervals, only point sizes. 

For example have a look at this [routing request](https://graphhopper.com/api/1/route?point=48.533736%2C9.192579&point=48.533722%2C9.192628&type=json&locale=en-AU&vehicle=car&weighting=fastest&elevation=true&key=2a24e316-61ea-4850-b231-4ef2fe25d229&points_encoded=false).

It has only 2 points. The instructions look like this:
```
instructions: [
{
distance: 3.928,
sign: 0,
interval: [
0,
1
],
text: "Continue onto K 6720",
time: 220,
street_name: "K 6720"
},
{
distance: 0,
sign: 4,
interval: [
1,
1
],
text: "Arrive at destination",
time: 0,
street_name: ""
}
]
```

The continue instruction has 1 point. The finish instruction has 1 point. The second number of the interval points to the next instruction. Example [0,1], here the next instruction would start at the point with the index 1. 

This is the current state of the API.

When thinking more about this, this actually seems wrong. I think the second number of the interval should point to the last point of the instruction.

This was also pointed out by @karussell [here](https://github.com/graphhopper/graphhopper/pull/1138#discussion_r135384510)

> All path details currently work this way: you specify first and last and the length is last-first.

**I finally understood the issue**.

Think about this turn:

```
A ----- B
         |
         C
```

If you route from A->C. The PathDetails for the street name would be:

```
details: {
street_name: [
[
0,
1,
"A-B"
],
[
1,
2,
"B-C"
]
]
}
```

From point 0 (A) to point 1 (B). We are on the street A-B. From point 1 (B) to point 2 (C) we are on the street "B-C". So the path details are correct.

For the same situation the instructions right now, would look like:
```
instructions: [
{
distance: 9.613,
sign: 0,
interval: [
0,
1
],
text: "Continue onto A-B",
time: 692
},
{
distance: 7,
sign: 2,
interval: [
1,
2
],
text: "Turn right onto B-C",
time: 280
},
{
distance: 0,
sign: 4,
interval: [
2,
2
],
text: "Arrive at destination",
time: 0,
street_name: ""
}
]
```

Which is also correct. The problem of the instructions are the format of the instructions, as instructions contain points and no index references. When we calculate the points of a route, we don't duplicate the points at the end/beginning of an edge. Therefore we have to use "hacks" in the serialization to artificially set points of some to size 0. 

I think we should change the format of the instructions to become more similar to the PathDetails, by not containing points, but only from and to references.

WDYT?

**Comment by devemux86** [comment]
In both cases I mean the Java clients, [GraphHopper](https://github.com/graphhopper/graphhopper/blob/master/core/src/main/java/com/graphhopper/GraphHopper.java) from `graphhopper-core` using offline graph and [GraphHopperWeb](https://github.com/graphhopper/directions-api-java-client/blob/master/client-hc/src/main/java/com/graphhopper/api/GraphHopperWeb.java) from `directions-api-client-hc` using an online server (tested GraphHopper and Kurviger the same).

Sorry when mentioned "intervals" above, it was the Java instruction point sizes. :slightly_smiling_face: 
Iterating `instruction.getPoints().size()` returns inconsistent results between the two clients for the exact same route.
The `directions-api-client-hc` client has +1 point sizes in the response instructions than the `graphhoppcer-core`.

**Comment by devemux86** [comment]
With `directions-api-client-hc` client, each instruction (except the final) has in its `PointList` as last point also the 1st point of the next instruction.
This is not happening with `graphhoppcer-core` client, where the instructions `PointList` end at previous point from the next instruction.

**Comment by karussell** [comment]
> Iterating instruction.getPoints().size() returns inconsistent results between the two clients for the exact same route.

Yeah, this could be. But I think the current Java API for gh-core is wrong and not the hc client.

> I think we should change the format of the instructions to become more similar to the PathDetails, by not containing points, but only from and to references.

Yes, this would be perfect (including the problem with the PointList that @devemux86 found) but we can also do this as a later step if too much effort. As a first step we can fix the interval 'bug' of the ViaInstructions.
